File: SistemaEstoqueMercado/frontend/services/localizacaoService.ts
import api from './api';
import { Localizacao } from '@/types';

export const localizacaoService = {
  listarTodos: async (): Promise<Localizacao[]> => {
    const response = await api.get<Localizacao[]>('/localizacoes');
    return response.data;
  },

  buscarPorId: async (id: number): Promise<Localizacao> => {
    const response = await api.get<Localizacao>(`/localizacoes/${id}`);
    return response.data;
  },

  cadastrar: async (localizacao: Localizacao): Promise<Localizacao> => {
    const response = await api.post<Localizacao>('/localizacoes', localizacao);
    return response.data;
  },

  atualizar: async (id: number, localizacao: Localizacao): Promise<Localizacao> => {
    const response = await api.put<Localizacao>(`/localizacoes/${id}`, localizacao);
    return response.data;
  },

  excluir: async (id: number): Promise<void> => {
    await api.delete(`/localizacoes/${id}`);
  },
};

File: SistemaEstoqueMercado/backend/src/main/java/j0aod3v/sistemaestoquemercado/service/LocalizacaoService.java
package j0aod3v.sistemaestoquemercado.service;

import j0aod3v.sistemaestoquemercado.model.Localizacao;
import j0aod3v.sistemaestoquemercado.repository.LocalizacaoRepository;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class LocalizacaoService {

    private final LocalizacaoRepository localizacaoRepository;

    public LocalizacaoService(LocalizacaoRepository localizacaoRepository) {
        this.localizacaoRepository = localizacaoRepository;
    }

    public Localizacao cadastrar(Localizacao localizacao) {
        validar(localizacao);
        return localizacaoRepository.save(localizacao);
    }

    public Localizacao buscarPorId(int id) {
        validarId(id);
        return localizacaoRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Localização não encontrada para o ID: " + id));
    }

    public List<Localizacao> listarTodos() {
        return localizacaoRepository.findAll();
    }

    public Localizacao atualizar(Localizacao localizacao) {
        validar(localizacao);
        validarId(localizacao.getIdLocalizacao());
        buscarPorId(localizacao.getIdLocalizacao());
        return localizacaoRepository.save(localizacao);
    }

    public void excluir(int id) {
        validarId(id);
        localizacaoRepository.deleteById(id);
    }

    private void validar(Localizacao localizacao) {
        if (localizacao == null) {
            throw new IllegalArgumentException("Localização não pode ser nula.");
        }
        if (localizacao.getCodigoPosicao() == null || localizacao.getCodigoPosicao().isBlank()) {
            throw new IllegalArgumentException("Código da posição é obrigatório.");
        }
        if (localizacao.getTipoArmazenamento() == null || localizacao.getTipoArmazenamento().isBlank()) {
            throw new IllegalArgumentException("Tipo de armazenamento é obrigatório.");
        }
    }

    private void validarId(int id) {
        if (id <= 0) {
            throw new IllegalArgumentException("ID inválido.");
        }
    }
}